### ISSUE TITLE
ScribusGenerator to recognise filepaths as vars and then trigger import of assets from external resources.

### ISSUE THREAD

**Issue by datavectors** [berteh/ScribusGenerator#167]
I have successfully used ScribusGenerator (SG/Python 3) on Ubuntu 20.04.

I now want to integrate SG into a workflow where variables are not just short strings or integers but include filepaths to external assets such as long blocks of text which can be imported. I remember reading somewhere about "longtext" but I intend to pull in large blocks of text which cannot go into a csv cell. However filepaths can.

Where do you suggest I break into the SG code to ..

- check if a csvData element is a filepath string
- read the remote asset
- write long text blocks into Scribus document (textframes) as like other SG vars.

Break point to call a helper python script for this import seems to be somewhere after line 142 #generation

...

To explain further.
I aim to keep content production (Atom) separate from layout (Scribus).

Although I can write text directly into textframes using Scribus GUI my plan is to create textblocks externally in Atom editor and write the absolute filepaths to these assets into the csv file. Then SG will be called through command line from Atom through `process-palette` package. 
In other words a toolchain.

[P.S.] I have read this [closed issue on SVG import](https://github.com/berteh/ScribusGenerator/issues/164) and I offered an idea I might now use for such "non text" imports.

**Comment by berteh** [comment]
I would personally be wary about automatically turning all filepath-like CSV into loading external assets... as you may want such values to be displayed as text as well in other cases (eg with a user guide)

How about rather establishing a convention on the VAR name that would trigger this behaviour instead?

For instance, with a CSV commun named 'file', %VAR_file% would display the file path, and %VAR_LOAD_file% would parse the file at the given address.

I like this approach more as it seems to not impact the usual generator behaviour for other users.


If that works for your case, then you could have a look at extending the table of substitutions (replacements) computed in 'substituteData': https://github.com/berteh/ScribusGenerator/blob/ScribusGeneratorBackend.py#L491
Detecting the variables with a given keyword (_LOAD_ in my example above), and creating and extra table entry with the variable name and file content... the rest should be automatic unless I miss smthg.

You'd need to handle missing file case (would the path displayed instead for instance?), access rights and maybe some border cases (Scribus errors in using the file content?)... but I see no reason why that would not work.

Good luck!
Kindly update the README accordingly in your choice push

**Comment by datavectors** [comment]
Thank you. The naming convention you propose seems a good plan to add to default VAR behaviour.
I have also read the problems about importing SVG (I added a P.S. to my initial post) and I will now think through how to use SG and UI automation in tandem.

**Comment by berteh** [comment]
That being said I can't imagine Scribus-compatible long text, as long as you'd imagine, that could not be included in a CSV file, provided proper escaping of line jumps and other brackets... just in case.

**Comment by datavectors** [comment]
I have approached this requirement by embedding PHP snippets as variables using SG.
Then in a second pass I use PHP local server to expand the SG embedded PHP snippets.
Credit to carto.net for the original idea. I can expand further if you wish. This seems a good spot to record it. A parallel approach I explored was driving Scribus UI through an automation script.

**Comment by berteh** [comment]
Glad you found a solution.

Yes indeed I'm interested in more, if you can post an example along with your php script that would be nice. attached files can simply be dropped here (sometimes adding a .txt suffix if the system rejects your extension). 

Afterwards we can copy-paste the text to the [wiki](https://github.com/berteh/ScribusGenerator/wiki) for  a bit more visibility.

**Comment by datavectors** [comment]
My experiments are at an early stage - work in progress - so I will take more time in tweaking.

Meanwhile, I post here the references I followed to arrive at my experimental workflow.
Basically I require to curate various external assets into a Scribus template document - dynamically - and then publish to either PDF or Web - or both.  My external tool is Atom editor which creates the raw assets which Scribus "includes" through PHP script (there is no PHP script to publish).

I have taken some ideas I first read some years ago in carto.net (now archived). 

This paper below refers to injecting PHP variables into SVG files - but then I realised that I can apply the same scripting process to SLA files which are also XML like SVG. Or even SVG template imported into SLA.  Dynamic SVG fonts for example using Raphael.js.

https://old.carto.net/papers/svg/samples/mysql_svg_php.shtml

and here ...

https://old.carto.net/papers/svg/samples/serverside_svg_php.shtml

However this is where it becomes quite tricky to follow. Note that my workflow is in Ubuntu 20.04 and Windows/Mac readers will have to adapt.

This archived example will not work in your browser since the archive links (showing dynamic changing of SVG elements on each refresh of browser) are broken .. but observe the blocks of PHP embedded in the code.  These php blocks use the short notation <? .. ?> but I suggest using the longer notation <?php .. ?>.

Now the second link gives a good explanation.

https://old.carto.net/papers/svg/samples/serverside_svg_php.shtml

Lines 01-06 are the required PHP header and a list of variables to be expanded. In this case SVG attributes width and height.

Lines 08-19 contain SVG source code which use variables $svgwidth and $svgheight.

The final point to note is that the second pass of SLA document requires the document to have extension *.pxml and loaded through PHP local server > view source,

For this, as written in the references, you have to go to /etc/mime.types

Where you see see .svg add the custom mime type .psvg.

Do the same for .xml adding .pxml.

image/svg+xml					svg svgz psvg
application/xml					xml xsd pxml

A PHP header in the document *.xml will be required for each of these types.

The next hack is to go to /etc/apache2/apache2.conf and

AddType application/x-httpd-php .php .phtml .psvg .pxml

The bottom line is that this second stage translation pass allows PHP variables to be expanded in Scribus SLA (document extension first changed from *.sla to *.pxml).

Equally the process can be run on SVG files, either alone or (to be tested) imported into SLA document.

As another refinement I am defining variables as json format and then converting json to csv to be processed by ScrbusGenerator, Using the json method I can introduce long strings such as filepaths for files/assets to be "included".

One point to note is that a PHP header is required at the top of the *.pxml document (aka *.sla). In that header I define all the variables in a block. Then the embedded snippets (placed in textframes in *.pxml) can simply be ..

<?php echo $chapter01; ?>
<?php echo $chapter02; ?>
...

where contents of $chapter01 and $chapter02 are first defined in the PHP header by loading external files..

It is a rather complicated workflow to setup. The aim is to emulate some of the features seen in InDesign. For example I can introduce footnotes in Atom markdown blocks to be reflected into the final document.
